stub_test.go

```diff
@@ -0,0 +1,547 @@
+			c.WriteBulkString(args[0])
+		} else {
+			c.WriteSimpleString("PONG")
+		}
+	})
+
+	// We register a default QUIT command handler
+	rs.RegisterCommandHandler("QUIT", func(c *Connection, args []string) {
+		c.WriteOK()
+		c.Close()
+	})
+
+	return nil
+}
+
+// Close stops the RedisStub server.
+func (rs *StubServer) Close() {
+	rs.Lock()
+	if rs.listener != nil {
+		err := rs.listener.Close()
+		if err != nil {
+			// this is unrecoverable, so we panic.
+			panic(err)
+		}
+	}
+	rs.listener = nil
+	rs.Unlock()
+
+	rs.waitGroup.Wait()
+}
+
+// RegisterCommandHandler registers a handler for a redis command.
+//
+// The handler is called when the command is received. It gives access
+// to a `*Connection` object, which can be used to send responses, using
+// its `Write*` methods.
+func (rs *StubServer) RegisterCommandHandler(command string, handler func(*Connection, []string)) {
+	rs.Lock()
+	defer rs.Unlock()
+	rs.handlers[strings.ToUpper(command)] = handler
+}
+
+// Addr returns the address of the RedisStub server.
+func (rs *StubServer) Addr() string {
+	return rs.boundAddr.String()
+}
+
+// HandledCommandsCount returns the total number of commands
+// ran since the redis stub server started.
+func (rs *StubServer) HandledCommandsCount() int {
+	rs.Lock()
+	defer rs.Unlock()
+	return rs.processedCommands
+}
+
+// HandledConnectionsCount returns the number of established client
+// connections since the redis stub server started.
+func (rs *StubServer) HandledConnectionsCount() int {
+	rs.Lock()
+	defer rs.Unlock()
+	return rs.connectionCount
+}
+
+// GotCommands returns the commands handled (ordered by arrival) by the redis server
+// since it started.
+func (rs *StubServer) GotCommands() [][]string {
+	rs.Lock()
+	defer rs.Unlock()
+	return rs.commandsHistory
+}
+
+// listenAndServe listens on the redis server's listener,
+// and handles client connections.
+func (rs *StubServer) listenAndServe(l net.Listener) {
+	for {
+		nc, err := l.Accept()
+		if err != nil {
+			return
+		}
+
+		rs.waitGroup.Add(1)
+		rs.Lock()
+		rs.connections[nc] = struct{}{}
+		rs.connectionCount++
+		rs.Unlock()
+
+		go func() {
+			defer rs.waitGroup.Done()
+			defer nc.Close() //nolint:errcheck
+
+			rs.handleConnection(nc)
+
+			rs.Lock()
+			delete(rs.connections, nc)
+			rs.Unlock()
+		}()
+	}
+}
+
+// handleConnection handles a single redis client connection.
+func (rs *StubServer) handleConnection(nc net.Conn) {
+	connection := NewConnection(bufio.NewReader(nc), bufio.NewWriter(nc))
+
+	for {
+		command, args, err := connection.ParseRequest()
+		if err != nil {
+			connection.WriteError(ErrInvalidSyntax)
+			return
+		}
+
+		rs.Lock()
+		request := append([]string{command}, args...)
+		rs.commandsHistory = append(rs.commandsHistory, request)
+		rs.Unlock()
+
+		rs.handleCommand(connection, command, args)
+		connection.Flush()
+
+		rs.Lock()
+		closed := connection.closed
+		rs.Unlock()
+		if closed {
+			nc.Close() //nolint:errcheck,gosec
+		}
```

Comment:
the caller is passing an open connection so it's already managing `net.Conn`, if we can leave this responsibility centralized I think would be better. `listenAndServe` is already calling defer `nc.Close()` so I'm just for removing this block.

Reply:
Good point indeed!